dialogue-core/src/main/java/com/palantir/dialogue/core/Balanced.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.codahale.metrics.SlidingTimeWindowArrayReservoir;
+import com.github.benmanes.caffeine.cache.Ticker;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.UnsafeArg;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Chooses nodes based on stats about each channel, i.e. how many requests are currently
+ * being served and also how many failures have been seen in the last few seconds.
+ *
+ * This is intended to be a strict improvement over Round Robin and Random Selection which can leave fast servers
+ * underutilized, as it sends the same number to both a slow and fast node. It is *not* appropriate for transactional
+ * workloads (where n requests must all land on the same server) or scenarios where cache warming is very important.
+ * {@link PinUntilErrorChannel} remains the best choice for these.
+ */
+final class Balanced implements LimitedChannel {
+    private static final Logger log = LoggerFactory.getLogger(Balanced.class);
+
+    /**
+     * We'd like to remember failures for a long time, but this would increase CPU on a hot codepath as we compute
+     * the number of failures for each channel in order to sort. See {@link ChannelWithStats#score}.
+     */
+    private static final Duration FAILURE_MEMORY = Duration.ofSeconds(5);
```

Comment:
We could use a simple single-bucket exponential decay implementation instead of sliding window. That way we can remember failures for a long time and simply cut the value in half every N seconds.

Reply:
Honestly given how well the sliding window performs in the simulations, I think I find it a little easier to reason about...